pkg/cmd/roachtest/operations/backup_restore.go

```diff
@@ -42,119 +41,224 @@ func (cl *backupRestoreCleanup) Cleanup(
 
-	if validate {
-		o.Status(fmt.Sprintf("verifying db %s matches %s", dbName, restoreDBName))
-		sourceFingerprints, err := fingerprintutils.FingerprintDatabase(ctx, conn, dbName, fingerprintutils.AOST(backupTS), fingerprintutils.Stripped())
-		if err != nil {
-			o.Fatal(err)
+	if !reflect.DeepEqual(sourceFingerprints, destFingerprints) {
+		b.operation.Fatalf("backup and restore fingerprints do not match: %v != %v", sourceFingerprints, destFingerprints)
+	}
+}
+
+func (b *backupRestoreOperation) scatter(ctx context.Context, dbName string) {
+	var tableNames []string
+
+	b.operation.Status(fmt.Sprintf("scattering database '%s'", dbName))
+
+	rows, err := b.conn.QueryContext(ctx, fmt.Sprintf("SELECT table_name FROM [SHOW TABLES FROM %s]", dbName))
+	if err != nil {
+		b.operation.Fatal(err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var tableName string
+		if err := rows.Scan(&tableName); err != nil {
+			b.operation.Fatal(err)
 		}
+		tableNames = append(tableNames, tableName)
+	}
+	if rows.Err() != nil {
+		b.operation.Fatal(err)
+	}
 
-		// No AOST here; the timestamps are rewritten on restore. But nobody else is touching this database, so that's fine.
-		destFingerprints, err := fingerprintutils.FingerprintDatabase(ctx, conn, restoreDBName, fingerprintutils.Stripped())
+	for _, table := range tableNames {
+		b.operation.Status(fmt.Sprintf("scattering table '%s.%s'", dbName, table))
+		_, err := b.conn.Exec(fmt.Sprintf("ALTER TABLE %s.%s SCATTER", dbName, table))
 		if err != nil {
-			o.Fatal(err)
+			b.operation.Fatal(err)
 		}
+	}
+}
 
-		if !reflect.DeepEqual(sourceFingerprints, destFingerprints) {
-			o.Fatalf("backup and restore fingerprints do not match: %v != %v", sourceFingerprints, destFingerprints)
-		}
+func (b *backupRestoreOperation) RunOfflineRestore(ctx context.Context, skipValidating bool) {
+	dbName := b.pickDB(ctx)
+	backup := b.backup(ctx, dbName, 0)
+	b.restoreOffline(ctx, backup, b.restoreDB)
+	if !skipValidating {
+		b.validateRestore(ctx, backup, b.restoreDB)
+	}
+}
+
+func (b *backupRestoreOperation) RunOnlineRestore(ctx context.Context, skipValidating bool) {
+	dbName := b.pickDB(ctx)
+	backup := b.backup(ctx, dbName, 0)
+
+	downloadJob := b.restoreOnline(ctx, backup, b.restoreDB)
+	if err := tests.WaitForSucceeded(ctx, b.conn, downloadJob, 8760*time.Hour /* 1 year - test specs define their own timeouts */); err != nil {
+		b.operation.Fatal(err)
 	}
 
-	return &backupRestoreCleanup{db: restoreDBName}
+	if !skipValidating {
+		b.validateRestore(ctx, backup, b.restoreDB)
+	}
 }
 
-func runBackupRestoreFn(
-	online bool, validate bool,
-) func(context.Context, operation.Operation, cluster.Cluster) registry.OperationCleanup {
-	return func(ctx context.Context, o operation.Operation, c cluster.Cluster) registry.OperationCleanup {
-		return runBackupRestore(ctx, o, c, online, validate)
+func (b *backupRestoreOperation) RunPartialRestore(ctx context.Context) {
+	const (
+		scatterWait  = 15 * time.Minute
```

Comment:
How did you choose a download wait of 15 minutes? For the `downloadWait`, we could instead wait until the download job has a fraction completed that we like.

What's the purpose of scatterWait? i believe ALTER TABLE SCATTER only returns once the relocations have completed, so i don't understand the purpose of waiting here.